src/deluge/gui/menu_item/arpeggiator/midi_cv/bass_probability.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2014-2023 Synthstrom Audible Limited
+ *
+ * This file is part of The Synthstrom Audible Deluge Firmware.
+ *
+ * The Synthstrom Audible Deluge Firmware is free software: you can redistribute it and/or modify it under the
+ * terms of the GNU General Public License as published by the Free Software Foundation,
+ * either version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program.
+ * If not, see <https://www.gnu.org/licenses/>.
+ */
+#pragma once
+#include "definitions_cxx.hpp"
+#include "gui/menu_item/integer.h"
+#include "gui/menu_item/value_scaling.h"
+#include "gui/ui/sound_editor.h"
+#include "model/clip/clip.h"
+#include "model/clip/instrument_clip.h"
+#include "model/drum/non_audio_drum.h"
+#include "model/instrument/kit.h"
+#include "model/song/song.h"
+
+namespace deluge::gui::menu_item::arpeggiator::midi_cv {
+class BassProbability final : public Integer {
+public:
+	using Integer::Integer;
+	void readCurrentValue() override {
+		Clip* currentClip = getCurrentInstrumentClip();
+		if (currentClip->output->type == OutputType::KIT) {
+			Drum* currentDrum = ((Kit*)currentClip->output)->selectedDrum;
+			if (currentDrum != nullptr
+			    && (currentDrum->type == DrumType::MIDI || currentDrum->type == DrumType::GATE)) {
+				auto* nonAudioDrum = (NonAudioDrum*)currentDrum;
+				this->setValue(computeCurrentValueForUnsignedMenuItem(nonAudioDrum->arpSettings.bassProbability));
+			}
+		}
+		else if (currentClip->type == ClipType::INSTRUMENT) {
+			this->setValue(
+			    computeCurrentValueForUnsignedMenuItem(((InstrumentClip*)currentClip)->arpSettings.bassProbability));
+		}
+	}
+
+	void writeCurrentValue() override {
+		int32_t value = computeFinalValueForUnsignedMenuItem(this->getValue());
+		Clip* currentClip = getCurrentClip();
+		if (currentClip->output->type == OutputType::KIT) {
+			Drum* currentDrum = ((Kit*)currentClip->output)->selectedDrum;
+			if (currentDrum != nullptr
+			    && (currentDrum->type == DrumType::MIDI || currentDrum->type == DrumType::GATE)) {
+				auto* nonAudioDrum = (NonAudioDrum*)currentDrum;
+				nonAudioDrum->arpSettings.bassProbability = value;
+			}
+		}
+		else if (currentClip->type == ClipType::INSTRUMENT) {
+			((InstrumentClip*)currentClip)->arpSettings.bassProbability = value;
+		}
```

Comment:
For the menu items, could all ARP menu items share a `beginSession` implementation that grabs the arp settings for the current context (so basically this logic), and stuffs that into `soundEdtior` somewhere? I think it would dramatically simplify the `writeCurrentValue`/`readCurrentValue` for the majority of these new menu items.